Add automated, regular binary builds

Kjarrigan opened the pull request:
### Summary

Weekly automated builds of binary packages for Windows, Linux and OSX

### What this PR does

I found a couple of tutorials to build packages for Windows, OSX and Linux through Travis. I want to suggest doing this on a regular basis (there are cronjobs available although in a Beta Status) to make sure someone can play/test without the need to install and have knownledge of unity. This way we could widen the audience and maybe get some feedback / bugreports from non-devs.

I already tried some stuff but the git log is actually a mess because I mainly edited/added files through the github interface. But if this idea is accepted I'll rebase the whole thing to one / a couple logical grouped commits.

> To check whether a build was triggered by cron, examine the TRAVIS_EVENT_TYPE environment variable to see if it has the value cron.

So there is also no need to run the build on every PR but just through the cron-job. So the CI-checks doesn't take longer than necessary.

### TODO List for me

- [x] Clean up the mess
- [x] Add a script to update the bintray.json
- [x] Discuss if bintray.com is the right place to store the packages

### TODO List for the Merger

#### Bintray

- [x] Register at [Bintray](https://bintray.com/) (can be done through the Github-Account)
- [x] Create an Organisation
- [x] Create an Repository

#### Code

- [x] Change deployment-Section of .travis.yml to the new Bintray-Account
- [x] Change the package-Section of ./Sripts/bintray.json.erb to the new Bintray-Account

#### Travis
- [ ] Configure an cron job -> [Travis-Cronjobs](https://docs.travis-ci.com/user/cron-jobs/)

### Links

[Setup Unitybuilds on Travis](http://blog.stablekernel.com/continuous-integration-for-unity-5-using-travisci)

[Travis-Deployment to Bintray](https://docs.travis-ci.com/user/deployment/bintray/)
[Downloads](https://bintray.com/kjarrigan/ProjectPorcupine/Prebuild-Binarys#files)

BraedonWooding commented:
Looks great so far, it's a brilliant idea!

koosemose commented:
I like the general premise, though I will note in its current state, it's not much of a "playable" thing... too many core systems still need work, but that's no reason to put off having something like this in place. Once you get the chance to clean it up a bit more, I'll see about looking at it in more detail.

Kjarrigan commented:
Thanks for your feedback. I'll prepare everything for further discussion and report back!

Kjarrigan commented:
Ok, I'm ready for more detailed review of my suggestion. I cleaned up the code. Its actually not much Todo. I ...
- ... changed the `Install/unity.sh `to add the platform building packages on demand.
- ... added an `build.sh`-Skript to create the 3 packages and zip them
- ... added an deployment-Section to `.travis.yml` to deploy to Bintray.com
- ... added an `bintray.json.erb` config file template which is _compiled_ in `build.sh`

I extended my description above and included a list of Todos for the merger, because obviously the login-Data for Bintray and the Target-Directory shouldn't be at my personal storage. I added some comments to highlight what has to be changed.

> All checks have passed

That's only half of the truth - it just says that I didn't broke anything. The confirmation that the cronjob / upload-Part worked can be seen here: 
https://travis-ci.org/Kjarrigan/ProjectPorcupine/builds/221728467

The stylecop only _failed_ because he tried to _deploy_ something without a config file (as there is really nothing to do). That's an easy fix - just another if-clause to only run the deployment for the unit-test. Altough I thought about making a new JOB called "build". I think that would be more readable? But thats just a minor stylistic change.

The cron-job in Travis currently runs the first time when u add a cronjob and then every timespan chosen again. So if you setup the cronjob at Wednesday 14:00 it'll run every week on Wednesday at about 14:00.

Kjarrigan commented on .travis.yml:
```diff
@@ -18,3 +18,12 @@ script:
+deploy:
+  provider: bintray
+  file: ./Scripts/bintray.json
+  user: Kjarrigan
```
Change to organization name

Kjarrigan commented on .travis.yml:
```diff
@@ -18,3 +18,12 @@ script:
+    secure: SNg/fU847/VkYR9Ta55Vy3w9WLbGbEjhl8O1JocPeEqJ+qqw0ztzde5p+kVFN22SqMBOtLJYMtPmgpyZlWeWlKqGpflvajD1BcF7ycMp1PAAv6FvLijnzcSmQMMxhon7J7bIHc7xn9gmSRGEW+IkaL+mjmyRZTnuoMjmwlLPmQBBwZPNxa3mAHqL7jltJsZVHQ6puHoe3qo0mlZBfzCaMHpsMfHlzRDTcAxq1HCYrraIQmCdHZ0niKwGJNMVzq9wA5puVcVxh95UuPUaf6Jbs3u8sGLIrGJdnfKn1mo3E/FP5ZLtDAB7gIqW1j3vXDYYFWKezyi5UT8ZW7htJnOVg3uKB/Y8zhFieQXTZs42jnw1v6pi40qtg9+E2c2UAVyaEub ...
+  on:
+    condition: $TRAVIS_EVENT_TYPE = cron
+    all_branches: true
```
Remove to only build master

BraedonWooding replied:
Remove all these 4 lines or only the `all_branches: true` bit?

Kjarrigan replied:
Just the all_branches: true. The "condition" parts tells travis to only deploy when the cron runs (so once a week) instead of running on every CI run, which would slow down the checks quite a bit.

Kjarrigan commented on .travis.yml:
```diff
@@ -18,3 +18,12 @@ script:
+  file: ./Scripts/bintray.json
+  user: Kjarrigan
+  key:
+    secure: SNg/fU847/VkYR9Ta55Vy3w9WLbGbEjhl8O1JocPeEqJ+qqw0ztzde5p+kVFN22SqMBOtLJYMtPmgpyZlWeWlKqGpflvajD1BcF7ycMp1PAAv6FvLijnzcSmQMMxhon7J7bIHc7xn9gmSRGEW+IkaL+mjmyRZTnuoMjmwlLPmQBBwZPNxa3mAHqL7jltJsZVHQ6puHoe3qo0mlZBfzCaMHpsMfHlzRDTcAxq1HCYrraIQmCdHZ0niKwGJNMVzq9wA5puVcVxh95UuPUaf6Jbs3u8sGLIrGJdnfKn1mo3E/FP5ZLtDAB7gIqW1j3vXDYYFWKezyi5UT8ZW7htJnOVg3uKB/Y8zhFieQXTZs42jnw1v6pi40qtg9+E2c2UAVyaEub ...
```
Change to organization API-KEY. Should be encrypted using 
`$ travis encrypt BINTRAY-API-KEY --add deploy.key`

Kjarrigan commented on Scripts/bintray.json.erb:
```diff
@@ -0,0 +1,22 @@
+<% version = Time.now.strftime("%Y.%V") # (year).(week), e.g. 2017.15 %>
+{
+    "package": {
```
Change all attributes to match the settings of the organization @ bintray.

Kjarrigan commented on Scripts/bintray.json.erb:
```diff
@@ -0,0 +1,22 @@
+<% version = Time.now.strftime("%Y.%V") # (year).(week), e.g. 2017.15 %>
```
As I'm a ruby dev (and all travis images have one already installed) I used erb as template to dynamically generate the json.

Kjarrigan commented on Scripts/bintray.json.erb:
```diff
@@ -0,0 +1,22 @@
+
+    "version": {
+        "name": "<%= version %>",
+        "released": "2017-04-13",
```
Damn - forgot to dynamically change the date. Will push an update soon.

BraedonWooding commented:
Okay, so I'm gonna let @koosemose give the final word, but as long as the bintray organisation is under OrderOfThePorcupine (which he probably needs to setup, and the reason being to keep everything under the same roof the reason why we moved), along with having the 'VCS' link to `https://github.com/OrderOfThePorcupine/ProjectPorcupine`, and all the changes you proposed above are executed (which from what I can gather require the organisation to be setup which again I'll let koose do it since he is the 'manager'/ceo or whatever you would call it off PP.  Good job, nice and clean code thumbs up from me !

koosemose commented:
My knowledge on this side of things is lacking, so most any opinions I could state on the matter would be wildly uninformed and serve little purpose other than some silly attempt to exert power, so my input will be minimal, aside from trying to get some input from contributors more experienced in such thing.

So once everything is completed to your satisfaction let me know, and let me know what and where things need set up, and what information you need for use of the scripts, and I'll get things set up.

BraedonWooding commented:
@Kjarrigan I fixed the conflict for you since I implemented the new version of unity, as you can see I just changed the version to the new version which is 5.5.3f1.  Other than that I didn't change anything.  Tell us when it is ready to be merged and for us to make the changes to put it under OrderOfThePorcupine!

Kjarrigan commented:
Hi @BraedonWooding, usually I got no time on weekends for coding. I'll polish the last remaining bits tomorrow and extend my Todo-List for @koosemose to give a little further explanation on what to do / set up.

Kjarrigan commented:
So I rebased against master (to only relevant part was to set the version to 5.5.3f1)

@koosemose I guess I'm done with my part as far as I can tell. If you do the Bintray stuff on my Todolist above I could change most of the code so you only have to do the last few bits. Bintray does a great job of guiding you through the whole creation process. So you shouldn't have any problems setting up the upload environment.

Step 1) Sign Up - https://bintray.com/signup/oss
Step 2) Click on "Add Organization" and follow the instructions
Step 3) Create an repository where the binaries should be stored (will be part of the step 2 process)

Report back and I'll prepare the rest. :smile:

koosemose commented:
Ok, got it set up (hopefully correctly), it's at https://bintray.com/orderoftheporcupine/ProjectPorcupine , what other information is needed for you to set this up?

Kjarrigan commented:
Looks good. I changed all parts so just the secure API-Key is missing. To get that...

1) Visit https://bintray.com/profile/edit and click on API-Key. 
2) Encrypt your API-Key. 

The [Travis Doc](https://docs.travis-ci.com/user/deployment/bintray/) says: 
> It is recommended that you encrypt your api key. You can encrypt this key using the `travis` command line client and this command:
> `$ travis encrypt BINTRAY-API-KEY --add deploy.key`

If your on a MAC just type `gem install travis` to get the CLI tool then cd in the ProjectPorcupine directory and call the travis command as shown above ([Travis Encryption Docs](https://docs.travis-ci.com/user/encryption-keys/))

It adds the key to your .travis.yml. Copy&Paste the `secure: KEY` and I'll insert it.

koosemose commented:
secure: PufaISwFWiFSLoBM/fMIBpABcTAmktf4pYS9nbEB5IrnYPm67FKLsFjVKofCkMVT/Z7/DoglqT88zit/g1nFklcAvAcQ3dz5zfWqftpIX5fWG7KnNggivnM0EZldj81lVlBuwQeg98F84VVK6qVAk1bcTxdosPn8gtGw3/QSGqWpMe1il38YFG8nb/hMx2lgJqw4s/55yau96N6/05WJdSbdmfdtptoz9oXRBr5Yq2vHwvtQPuNJNaxx7fCcf4URPrb78vShB+Uvj3+Lppvmmf7xIKhcSPXLIEXN8Szoa+CSBdbn6jm/7gD4mPdpN8UeK9MtqskSeg2lrdOw6ABzfxFmBfd+NlktTBE01uAXLJmvtoarcyR5jLAI+I0306YjmxYL4ZW7SwcSTaOLz6N6CvzbkAWSk6jJ51Uq90bpVE7k6d1VaA5V9DKdAjL1AdIuyMNMmal+wzo43/7jbij6iBBO482AjEkxEIsbS6C1UXMdfBDwRjT0Os3FcXXX3VVARiRXfXn7ZfYvFpvt1sSil6rxsZkHPrg9egvNLG2fjOVR79+cDQII0xGv25z8cNrVV9ligXCxjJrE4CW5EQ8Jp3188E9QDek/6dt3tgRUS4egA83o+aYLAeQDIYtc09U1gbw6SFI+QhLBllHO0WFPjaAm+MGJtewfwVYuAUNZmXE=

Kjarrigan commented:
Ok. I pushed a second commit which changes all values to the "official" ones. So i guess we're ready to merge. If it really works we just can find out by starting the cron-job after merge and (fingers crossed) see if something is uploaded to bintray.

Setting up an cron job is actually pretty easy. Just click on [Settings](https://travis-ci.org/OrderOfThePorcupine/ProjectPorcupine/settings) @ travis-ci.org and at the bottom of the page there is the "Cron Jobs" section. Change the interval to weekly and click `Add`? Done. Then just wait a few minutes until the jobs starts and then lets see what happens.

koosemose commented:
Here's hopin things don't break catastrophically...

koosemose commented:
And it errors
```
[Bintray Upload] Reading descriptor file: ./Scripts/bintray.json
/Users/travis/.rvm/gems/ruby-2.2.1/gems/dpl-1.8.31/lib/dpl/provider/bintray.rb:54:in `read': No such file or directory @ rb_sysopen - ./Scripts/bintray.json (Errno::ENOENT)
	from /Users/travis/.rvm/gems/ruby-2.2.1/gems/dpl-1.8.31/lib/dpl/provider/bintray.rb:54:in `read_descriptor'
	from /Users/travis/.rvm/gems/ruby-2.2.1/gems/dpl-1.8.31/lib/dpl/provider/bintray.rb:443:in `deploy'
	from /Users/travis/.rvm/gems/ruby-2.2.1/gems/dpl-1.8.31/lib/dpl/cli.rb:32:in `run'
	from /Users/travis/.rvm/gems/ruby-2.2.1/gems/dpl-1.8.31/lib/dpl/cli.rb:7:in `run'
	from /Users/travis/.rvm/gems/ruby-2.2.1/gems/dpl-1.8.31/bin/dpl:5:in `<top (required)>'
	from /Users/travis/.rvm/gems/ruby-2.2.1/bin/dpl:23:in `load'
	from /Users/travis/.rvm/gems/ruby-2.2.1/bin/dpl:23:in `<main>'
```
Also this error was located on the Stylecop job, that doesn't seem right as that job doesn't even download Unity. It should probably either hook into the Unit Test Job (as it actually uses unity) or have its own job.

Kjarrigan commented:
Damn. I missed these two. I'll provide a fix asap! 

for the stylecop Job we need a second deploy-condition
and the in the bintry.json.erb are missing double-quotes around the date.
